cmd/snap/cmd_run.go

```diff
@@ -1132,6 +1132,14 @@ func (x *cmdRun) runSnapConfine(info *snap.Info, securityTag, snapApp, hook stri
 		env["XAUTHORITY"] = xauthPath
 	}
 
+	// Guarantee that XDG_RUNTIME_DIR does exist before launching the snap
+	if xdg_runtime_dir, exists := env["XDG_RUNTIME_DIR"]; exists {
+		if err = os.Mkdir(xdg_runtime_dir, 0700); err != nil && err != os.ErrExist {
```

Comment:
```suggestion
		if err = os.Mkdir(xdg_runtime_dir, 0700); err != nil && !errors.Is(err, os.ErrExist) {
```

The `os.Mkdir` function returns an `os.PathError` wrapping the syscall error, so it's never going to return something equal to `os.ErrExist`. So we need a test that can check the underlying error.

Reply:
I added a check before to avoid using Mkdir if the folder already exists.

Reply:
I think the version without the Stat call was closer to correct: the new version seems like it could  lead to TOCTOU problems.

Imagine that I tried to start two instances of the snap app at the same time. Each instance does the `os.Stat` call and finds that the directory doesn't exist. They both move on to the `os.Mkdir` call: one succeeds, and the other gets EEXIST.

If we just do the `os.Mkdir` and ignore the EEXIST error, things should be fine. As for how to check the error, this short playground example demonstrates the difference between a direct equality check and using `errors.Is`: https://go.dev/play/p/380VKMm1t_9

Reply:
Mmm... interesting. Thanks!